src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -363,90 +366,106 @@ private void ExectueClusterTransactionTask()
                     return;
 
                 _hasClusterTransaction.Reset();
-
-                ExectueClusterTransactionOnDatabase();
-            }
-        }
-
-        public void ExectueClusterTransactionOnDatabase(Action<List<Task>> customWait = null)
-        {
-            try
-            {
-                var transactionsTasks = new List<Task>();
-                using (PreventFromUnloading())
-                using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
-                using (serverContext.OpenReadTransaction())
+               
+                using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                using (context.OpenReadTransaction())
                 {
-                    var commands = ClusterTransactionCommand.ReadCommandsBatch(serverContext, Name, fromCount: _clusterTxCount).ToList();
-                    if (commands.Count == 0)
+                    try
                     {
-                        return;
-                    }
+                        var batch = new List<ClusterTransactionCommand.SingleClusterDatabaseCommand>(
+                            ClusterTransactionCommand.ReadCommandsBatch(context, Name, fromCount: _nextClusterCommand));
 
-                    foreach (var command in commands)
-                    {
-                        var mergedCommands = new BatchHandler.ClusterTransactionMergedCommand(this, command);
-                        var task = TxMerger.Enqueue(mergedCommands);
-                        transactionsTasks.Add(task);
-                        OnClusterTransactionCompletion(command.Index, task, mergedCommands);
+                        if (batch.Count == 0)
+                        {
+                            continue;
+                        }
+                        
+                        try
+                        {
+                            var mergedCommands = new BatchHandler.ClusterTransactionMergedCommand(this, batch);
+                            TxMerger.Enqueue(mergedCommands).Wait(DatabaseShutdown);
+                            foreach (var command in batch)
```

Comment:
Only the `Wait` here (for the completion of the commands) need to be in the try catch for the one at a time run.